eoe/man/man1/fpmode.1
'\"macro stdmacro
.if n .pH g1.fpmode @(#)fpmode
.nr X
.if \nX=0 .ds x} FPMODE 1 "User Environment Utilities" "\&"
.if \nX=1 .ds x} FPMODE 1 "User Environment Utilities"
.if \nX=2 .ds x} FPMODE 1 "" "\&"
.if \nX=3 .ds x} FPMODE "" "" "\&"
.TH \*(x}
.tr ~
.SH NAME
fpmode \- run a command in specified floating point and/or memory mode
.SH SYNOPSIS
.B fpmode
precise|performance|smm|nsmm|spec|nonspec command [ arguments ]
.SH DESCRIPTION
.I fpmode\^
executes
.I command\^,
forcing it to use the specified floating point exception and memory access
mode. This is
only meaningful on processors that support multiple modes. On
these processors, the default mode is usually performance mode (also
called imprecise exception mode) and non-sequential memory (nsmm) mode.
On all other processors, the default is precise exception and sequential (smm)
mode.
.PP
The R8000 is the only processor which supports both a precise exception
mode and a performance mode.  All other processors support only precise
exceptions, though automatic flushing of denormalized results to zero
may be controlled through the setting of the FS bit in the FP
control and status register by calling \f2set_fpc_csr()\fP.
.PP
In precise exception mode, all floating point exceptions/interrupts are
reported on the instruction that caused the exception, so floating point
signal handlers can clean up the FP state and continue execution. In
performance mode, floating point interrupts are asynchronous and the
reported program counter is meaningless. Precise mode is provided as
a means of debugging processes and/or backward compatibility for programs
that have knowledge of the FP state. It causes significant performance
degradation.
.PP
When a program is run in performance mode on an R8000, the floating point processor
automatically flushes all denormalized results to zero regardless of the
setting of the FS (flush denormalized results to zero) bit in the FP
control and status register. This bit is turned on at \f2exec()\fP time.
Turning it off (by calling \f2set_fpc_csr()\fP) will cause performance
degradation. When the same program is run in precise mode, the FS bit is
turned off at \f2exec()\fP time (for backward compatibility with older
processors). Note that this means a given program may generate different
results when run in performance and precise modes.
.PP
Non-sequential memory access mode allows the processor to do
floating point and integer memory operations out of order, thus
possibly providing better performance.  This is only supported on R8000
processors.
.PP
Turning on speculative execution (spec) mode tells the kernel to
ignore all memory access faults generated by the program. This mode
is used by the compiler to provide better performance in some cases
by allowing eager instruction scheduling. Note that this may also cause
significantly worse performance if used indiscriminately and should
be avoided when debugging since it masks potential problems.
.PP
The floating point mode is inherited across \f2fork()\fP and
\f2exec()\fP system calls. To modify both exception and memory modes,
use fpmode twice; e.g.,
.sp
.RS
fpmode precise fpmode smm command
.RE
.sp
.SH SEE ALSO
syssgi(2), get_fpc_csr(3), set_fpc_csr(3).
.SH WARNINGS
In the case of the following command
.sp
.RS
fpmode precise command1; command2
.RE
.sp
\f2fpmode\f1
applies only to
command1.
The command
.sp
.RS
fpmode precise (command1; command2)
.RE
.sp
is syntactically incorrect.
.Ee
'\".so /pubs/tools/origin.att


eoe/cmd/fpmode/fpmode.c
/**************************************************************************
 *									  *
 * 		 Copyright (C) 1988, Silicon Graphics, Inc.		  *
 *									  *
 *  These coded instructions, statements, and computer programs  contain  *
 *  unpublished  proprietary  information of Silicon Graphics, Inc., and  *
 *  are protected by Federal copyright law.  They  may  not be disclosed  *
 *  to  third  parties  or copied or duplicated in any form, in whole or  *
 *  in part, without the prior written consent of Silicon Graphics, Inc.  *
 *									  *
 **************************************************************************/

#ident	"$Revision: 1.6 $"

/*
**	fpmode -- run a process in precise|performance|smm|nsmm|spec|nonspec mode
*/

#include	<stdio.h>
#include	<ctype.h>
#include	"sys/types.h"
#include	"sys/syssgi.h"

main(argc, argv)
int argc;
char *argv[];
{
	extern	errno;
	extern	char *sys_errlist[];
	int	precise = -1, smm = -1, spec = -1;
#ifdef NOTDEF
	int	softfp = -1;
#endif

	if (argc < 3) {
usage:
		fprintf(stderr,
			"usage: %s precise|performance|smm|nsmm|spec|nonspec command [args...]\n",
			argv[0]);
		exit(2);
	}

	if (strcmp(argv[1], "precise") == 0)
		precise = 1;
	else if (strcmp(argv[1], "performance") == 0)
		precise = 0;
	else if (strcmp(argv[1], "smm") == 0)
		smm = 1;
	else if (strcmp(argv[1], "nsmm") == 0)
		smm = 0;
	else if (strcmp(argv[1], "spec") == 0)
		spec = 1;
	else if (strcmp(argv[1], "nonspec") == 0)
		spec = 0;
#ifdef NOTDEF
	else if (strcmp(argv[1], "oldsoftfp") == 0)
		softfp = 1;
	else if (strcmp(argv[1], "newsoftfp") == 0)
		softfp = 0;
#endif
	else
		goto usage;

	if ((precise != -1) && (syssgi(SGI_SET_FP_PRECISE, precise) < 0)) {
		fprintf(stderr,
			"%s: could not SGI_SET_FP_PRECISE %d -- %s\n ",
			argv[0], precise, sys_errlist[errno]);
		exit(2);
	}

	if ((smm != -1) && (syssgi(SGI_SET_CONFIG_SMM, smm) < 0)) {
		fprintf(stderr,
			"%s: could not SGI_SET_CONFIG_SMM %d -- %s\n ",
			argv[0], smm, sys_errlist[errno]);
		exit(2);
	}

	if ((spec != -1) && (syssgi(SGI_SPECULATIVE_EXEC, spec) < 0)) {
		fprintf(stderr,
			"%s: could not SGI_SPECULATIVE_EXEC %d -- %s\n ",
			argv[0], spec, sys_errlist[errno]);
		exit(2);
	}

#ifdef NOTDEF
	if ((softfp != -1) && (syssgi(SGI_OLD_SOFTFP, softfp) < 0)) {
		fprintf(stderr,
			"%s: could not SGI_OLD_SOFTFP %d -- %s\n ",
			argv[0], softfp, sys_errlist[errno]);
		exit(2);
	}
#endif

	if (syssgi(SGI_SET_FP_PRESERVE, 1) < 0) {
		fprintf(stderr,
			"%s: could not SGI_SET_FP_PRESERVE -- %s\n ",
			argv[0], sys_errlist[errno]);
		exit(2);
	}

	execvp(argv[2], &argv[2]);
	fprintf(stderr, "%s: %s\n", sys_errlist[errno], argv[2]);
	exit(2);
}
